File: lim.zig
const std = @import("std");
const print = std.debug.print;

const Country = struct {
    name: []const u8,
    population: i32,
};

const countries = [_]Country {
    .{ .name = "China", .population = 1_412_600_000 },
    .{ .name = "India", .population = 1_375_586_000 },
    .{ .name = "United States", .population = 333_299_945 },
    .{ .name = "Indonesia", .population = 275_773_800 },
    .{ .name = "Pakistan", .population = 235_825_000 },
    .{ .name = "Nigeria", .population = 218_541_000 },
    .{ .name = "Brazil", .population = 215_424_390 },
    .{ .name = "Bangladesh", .population = 165_158_616 },
    .{ .name = "Russia", .population = 145_100_000 },
    .{ .name = "Mexico", .population = 128_533_664 },
};

pub fn main() void {
    // also interesting because arbitrary sizes
    var total: i32 = 0;
    for (countries) |country| {
        print("{s} {}\n", .{country.name, fmtIntSep(country.population)});
        total += country.population;
    }
    print("total {}\n", .{fmtIntSep(total)});
}

fn fmtIntSep(n: anytype) FmtIntSep {
    return .{
        .n = std.math.absCast(n),
        .negative = n < 0,
    };
}

const FmtIntSep = struct {
    n: u128,
    negative: bool,

    pub fn format(
        self: FmtIntSep,
        comptime fmt_str: []const u8,
        options: std.fmt.FormatOptions,
        writer: anytype,
    ) !void {
        _ = fmt_str;
        _ = options;
        // Count digits
        var x = self.n;
        var exp: @TypeOf(self.n) = 1;
        var digits: std.math.Log2Int(@TypeOf(self.n)) = 0;
        while (x > 0) {
            x /= 10;
            exp *= 10;
            digits += 1;
        }
        // Print integer
        if (digits == 0) {
            return writer.writeAll("0");
        }
        if (self.negative) {
            try writer.writeByte('-');
        }
        while (digits > 0) {
            exp /= 10;
            digits -= 1;
            try writer.writeByte('0' + @intCast(u8, self.n / exp % 10));
            if (digits > 0 and digits % 3 == 0) {
                try writer.writeByte(',');
            }
        }
    }
};

File: lim.rs
use num_format::{CustomFormat, ToFormattedString, WriteFormatted};
use std::fmt;

struct Country<'a> {
    name: &'a str,
    population: i32,
}

const COUNTRIES: &[Country] = &[
    Country { name: "China", population: 1_412_600_000 },
    Country { name: "India", population: 1_375_586_000 },
    Country { name: "United States", population: 333_299_945 },
    Country { name: "Indonesia", population: 275_773_800 },
    Country { name: "Pakistan", population: 235_825_000 },
    Country { name: "Nigeria", population: 218_541_000 },
    Country { name: "Brazil", population: 215_424_390 },
    Country { name: "Bangladesh", population: 165_158_616 },
    Country { name: "Russia", population: 145_100_000 },
    Country { name: "Mexico", population: 128_533_664 },
];

fn main() {
    let format = CustomFormat::builder().separator("_").build().unwrap();
    let mut total = 0;
    for country in COUNTRIES {
        println!("{} {}", country.name, IntSep(country.population, &format));
        total += country.population;
    }
    println!("total {}", IntSep(total, &format));
}

struct IntSep<'a, N>(N, &'a CustomFormat)
where
    N: ToFormattedString;

impl<'a, N> fmt::Display for IntSep<'a, N>
where
    N: ToFormattedString,
{
    fn fmt(&self, mut formatter: &mut fmt::Formatter) -> fmt::Result {
        formatter.write_formatted(&self.0, self.1).unwrap();
        Ok(())
    }
}

// [1] https://doc.rust-lang.org/std/#primitives
// [2] https://doc.rust-lang.org/std/primitive.u128.html
